README.md
# Ouroboros

Ouroboros is a command line utility that can be used to recreate template solutions. Solutions for the following frameworks are supported:

* .NET

## Usage

```
usage: ouroboros [--version] [--help] <command> [<args>]

Available commands are:
    dotnet     Create a new .NET solution from a .NET solution
    version    Print ouroboros version and quit
```

### dotnet

```
usage: ouroboros dotnet <args> [tpl path] [dst path]
  
Available args are:  
    new-guids     Generate new project GUIDs
    tpl-name      Override template solution name
```

## Install

To install, use `go get`:

```bash
$ go get -d github.com/jjafuller/ouroboros
```

## Contribution

1. Fork ([https://github.com/jjafuller/ouroboros/fork](https://github.com/jjafuller/ouroboros/fork))
1. Create a feature branch
1. Commit your changes
1. Rebase your local changes against the master branch
1. Run test suite with the `go test ./...` command and confirm that it passes
1. Run `gofmt -s`
1. Create a new Pull Request

## Author

[jjafuller](https://github.com/jjafuller)

CHANGELOG.md
## 0.1.1 (2016-12-12)

### Added

- Flag for .NET to override template name

### Fixed

- Ignore item filter improvements

## 0.1.0 (2016-11-19)

Initial release

### Added

- Add .NET support

### Deprecated

- Nothing

### Removed

- Nothing

### Fixed

- Nothing


command/dotnet_test.go
package command

import (
	"testing"

	"github.com/mitchellh/cli"
)

func TestDotnetCommand_implement(t *testing.T) {
	var _ cli.Command = &DotnetCommand{}
}

func TestDotnetCommand_extractGUIDSln(t *testing.T) {
	var command = &DotnetCommand{}

	contents := `
Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Test", "Test\Test.csproj", "{8EA60CA5-7D3D-4813-ACB1-069618285452}"
EndProject
Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Test.Tests", "Test.Tests\Test.Tests.csproj", "{E20EF42E-DBB7-42AF-B870-6B1C9D27FC48}"
EndProject
`

	command.ExtractGUID(contents, ".sln")
}

func TestDotnetCommand_extractGUIDProj(t *testing.T) {
	var command = &DotnetCommand{}

	contents := `
    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
    <ProjectGuid>{E20EF42E-DBB7-42AF-B870-6B1C9D27FC48}</ProjectGuid>
    <OutputType>Library</OutputType>
`

	command.ExtractGUID(contents, "csproj")
}


cli.go
package main

import (
	"fmt"
	"os"

	"github.com/jjafuller/ouroboros/command"
	"github.com/mitchellh/cli"
)

// Run is the entry point for the program
func Run(args []string) int {
	// Meta-option for executables.
	// It defines output color and its stdout/stderr stream.
	meta := &command.Meta{
		Ui: &cli.ColoredUi{
			InfoColor:  cli.UiColorBlue,
			ErrorColor: cli.UiColorRed,
			Ui: &cli.BasicUi{
				Writer:      os.Stdout,
				ErrorWriter: os.Stderr,
				Reader:      os.Stdin,
			},
		}}

	return RunCustom(args, Commands(meta))
}

// RunCustom is like Run, but allows for preprocessing some flags
func RunCustom(args []string, commands map[string]cli.CommandFactory) int {
	// Get the command line args. We shortcut "--version" and "-v" to
	// just show the version.
	for _, arg := range args {
		if arg == "-v" || arg == "-version" || arg == "--version" {
			newArgs := make([]string, len(args)+1)
			newArgs[0] = "version"
			copy(newArgs[1:], args)
			args = newArgs
			break
		}
	}

	cli := &cli.CLI{
		Args:       args,
		Commands:   commands,
		Version:    Version,
		HelpFunc:   cli.BasicHelpFunc(Name),
		HelpWriter: os.Stdout,
	}

	exitCode, err := cli.Run()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to execute: %s\n", err.Error())
	}

	return exitCode
}


command/dotnet/vars.go
package dotnet

// VsData contains visual studio structural data
// ProjectTypeGUID is a map of known Visual Studio project type GUIDs
// SourceFileExts is a map of known source type file extensions
// ProjectFileExts is a map of known project type file extensions
var VsData = struct {
	ProjectTypeGUID, SourceFileExts, ProjectFileExts, IgnoredItems map[string]string
}{
	ProjectTypeGUID: map[string]string{
		"{06A35CCD-C46D-44D5-987B-CF40FF872267}": "Deployment Merge Module",
		"{14822709-B5A1-4724-98CA-57A101D1B079}": "Workflow (C#)",
		"{20D4826A-C6FA-45DB-90F4-C717570B9F32}": "Legacy (2003) Smart Device (C#)",
		"{2150E333-8FDC-42A3-9474-1A3956D46DE8}": "Solution Folder",
		"{2DF5C3F4-5A5F-47a9-8E94-23B4456F55E2}": "XNA (XBox)",
		"{32F31D43-81CC-4C15-9DE6-3FC5453562B6}": "Workflow Foundation",
		"{349C5851-65DF-11DA-9384-00065B846F21}": "Web Application (incl. MVC 5)",
		"{3AC096D0-A1C2-E12C-1390-A8335801FDAB}": "Test",
		"{3D9AD99F-2412-4246-B90B-4EAA41C64699}": "Windows Communication Foundation (WCF)",
		"{3EA9E505-35AC-4774-B492-AD1749C4943A}": "Deployment Cab",
		"{4D628B5B-2FBC-4AA6-8C16-197242AEB884}": "Smart Device (C#)",
		"{4F174C21-8C12-11D0-8340-0000F80270F8}": "Database (other project types)",
		"{54435603-DBB4-11D2-8724-00A0C9A8B90C}": "Visual Studio 2015 Installer Project Extension",
		"{593B0543-81F6-4436-BA1E-4747859CAAE2}": "SharePoint (C#)",
		"{603C0E0B-DB56-11DC-BE95-000D561079B0}": "ASP.NET MVC 1.0",
		"{60DC8134-EBA5-43B8-BCC9-BB4BC16C2548}": "Windows Presentation Foundation (WPF)",
		"{68B1623D-7FB9-47D8-8664-7ECEA3297D4F}": "Smart Device (VB.NET)",
		"{66A26720-8FB5-11D2-AA7E-00C04F688DDE}": "Project Folders",
		"{6BC8ED88-2882-458C-8E55-DFD12B67127B}": "MonoTouch",
		"{6D335F3A-9D43-41b4-9D22-F6F17C4BE596}": "XNA (Windows)",
		"{76F1466A-8B6D-4E39-A767-685A06062A39}": "Windows Phone 8/8.1 Blank/Hub/Webview App",
		"{786C830F-07A1-408B-BD7F-6EE04809D6DB}": "Portable Class Library",
		"{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}": "ASP.NET 5",
		"{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}": "C++",
		"{978C614F-708E-4E1A-B201-565925725DBA}": "Deployment Setup",
		"{A1591282-1198-4647-A2B1-27E5FF5F6F3B}": "Silverlight",
		"{A5A43C5B-DE2A-4C0C-9213-0A381AF9435A}": "Universal Windows Class Library",
		"{A860303F-1F3F-4691-B57E-529FC101A107}": "Visual Studio Tools for Applications (VSTA)",
		"{A9ACE9BB-CECE-4E62-9AA4-C7E7C5BD2124}": "Database",
		"{AB322303-2255-48EF-A496-5904EB18DA55}": "Deployment Smart Device Cab",
		"{B69E3092-B931-443C-ABE7-7E7B65F2A37F}": "Micro Frmework",
		"{BAA0C2D2-18E2-41B9-852F-F413020CAA33}": "Visual Studio Tools for Office (VSTO)",
		"{BC8A1FFA-BEE3-4634-8014-F334798102B3}": "Windows Store Apps (Metro Apps)",
		"{BF6F8E12-879D-49E7-ADF0-5503146B24B8}": "C# in Dynamics 2012 AX AOT",
		"{C089C8C0-30E0-4E22-80C0-CE093F111A43}": "Windows Phone 8/8.1 App (C#)",
		"{C252FEB5-A946-4202-B1D4-9916A0590387}": "Visual Database Tools",
		"{CB4CE8C6-1BDB-4DC7-A4D3-65A1999772F8}": "Legacy (2003) Smart Device (VB.NET)",
		"{D399B71A-8929-442a-A9AC-8BEC78BB2433}": "XNA (Zune)",
		"{D59BE175-2ED0-4C54-BE3D-CDAA9F3214C8}": "Workflow (VB.NET)",
		"{DB03555F-0C8B-43BE-9FF9-57896B3C5E56}": "Windows Phone 8/8.1 App (VB.NET)",
		"{E24C65DC-7377-472B-9ABA-BC803B73C61A}": "Web Site",
		"{E3E379DF-F4C6-4180-9B81-6769533ABE47}": "ASP.NET MVC 4.0",
		"{E53F8FEA-EAE0-44A6-8774-FFD645390401}": "ASP.NET MVC 3.0",
		"{E6FDF86B-F3D1-11D4-8576-0002A516ECE8}": "J#",
		"{EC05E597-79D4-47f3-ADA0-324C4F7C7484}": "SharePoint (VB.NET)",
		"{EFBA0AD7-5A72-4C68-AF49-83D382785DCF}": "Xamarin.Android / Mono for Android",
		"{F135691A-BF7E-435D-8960-F99683D2D49C}": "Distributed System",
		"{F184B08F-C81C-45F6-A57F-5ABD9991F28F}": "VB.NET",
		"{F2A71F9B-5D33-465A-A702-920D77279786}": "F#",
		"{F5B4F3BC-B597-4E2B-B552-EF5D8A32436F}": "MonoTouch Binding",
		"{F85E285D-A4E0-4152-9332-AB1D724D3325}": "ASP.NET MVC 2.0",
		"{F8810EC1-6754-47FC-A15F-DFABD2E3FA90}": "SharePoint Workflow",
		"{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}": "C#",
	},
	SourceFileExts: map[string]string{
		".asmx":   "Deployment",
		".asp":    "Active Server Page",
		".c":      "C/C++ source",
		".ccp":    "C++ source",
		".cs":     "C# source",
		".disco":  "Dynamic discovery document (handles XML web service discovery)",
		".fs":     "F# source",
		".h":      "Header (include)",
		".java":   "Java source",
		".jsl":    "J# source",
		".vb":     "VB source",
		".config": "Config file",
	},
	ProjectFileExts: map[string]string{
		".atp":             "Application template project",
		".csproj":          "C# project",
		".dbp":             "Database project",
		".fsproj":          "F# project",
		".hxc":             "Help project",
		".md":              "Markdown",
		".sln":             "Solution",
		".suo":             "Solution options",
		".vap":             "Visual Studio Analyzer project",
		".vbg":             "Compatible project group",
		".vbp":             "Visual Basic project",
		".vip":             "Visual Basic project",
		".vbproj":          "Visual Basic project",
		".vcxproj":         "Visual C++ project",
		".vcxproj.filters": "Solution explorer add file filter",
		".vdproj":          "Visual Studio deployment project",
		".vmx":             "Macro project",
		".vup":             "Utility project",
	},
	IgnoredItems: map[string]string{
		".DS_Store": "macOS metadata",
		"Thumbs.db": "Windows thumbnail cache",
		".vs":       "Visual Studio user dir",
		".git":      "git repo",
		"bin":       "binary output",
		"obj":       "object output",
		"packages":  "nuget packages",
	}}


command/dotnet.go
package command

import (
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"path"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"github.com/jjafuller/ouroboros/command/dotnet"
	"github.com/satori/go.uuid"
)

// in solution
// ^Project.*{(.*)}"$
// Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Test", "Test\Test.csproj", "{8EA60CA5-7D3D-4813-ACB1-069618285452}"

// in project
// ^\s*<ProjectGuid>{(.*)}</ProjectGuid>\s*$
//     <ProjectGuid>{8EA60CA5-7D3D-4813-ACB1-069618285452}</ProjectGuid>

var (
	tplPath, tplName, dstPath, dstName string
	guidInSln                          = regexp.MustCompile(`^Project.*{(?P<guid>.*)}"$`)
)

// DotnetCommand creates a .NET solution from a .NET solution
type DotnetCommand struct {
	Meta
}

// Run executes this command
func (c *DotnetCommand) Run(args []string) int {
	//fmt.Printf("%v", args) // dump args

	cmdFlags := flag.NewFlagSet("dotnet", flag.ContinueOnError)
	cmdFlags.Usage = func() { c.Ui.Output(c.Help()) }

	newGUIDs := cmdFlags.Bool("new-guids", false, "generate new guids?")

	overrideTplName := cmdFlags.String("tpl-name", "", "override template solution name")

	if err := cmdFlags.Parse(args); err != nil {
		c.Ui.Error(fmt.Sprintf("Invalid arguments: %s\n", err))
		return 1
	}

	if len(cmdFlags.Args()) != 2 {
		c.Ui.Error("The dotnet command expects two arguments\n")
		return 1
	}

	tplPath = cmdFlags.Arg(0)
	dstPath = cmdFlags.Arg(1)

	if len(*overrideTplName) < 1 {
		tplName = filepath.Base(tplPath)
	} else {
		tplName = *overrideTplName
	}

	dstName = filepath.Base(dstPath)

	fmt.Printf("Template: \n  %s\n", tplName)
	fmt.Printf("Destination: \n  %s\n", dstName)

	// ensure destination directory
	if _, err := os.Stat(dstPath); os.IsNotExist(err) {
		os.Mkdir(dstPath, 0755)
	}

	dirList, slnPath, err := getDirectoryList(tplPath)

	if err != nil {
		fmt.Printf("getDirectoryList() returned %v\n", err)
		return 1
	}

	if len(slnPath) < 1 {
		fmt.Println("No solution file was found within the directory")
		return 1
	}

	fmt.Printf("Solution file: \n  %s\n", slnPath)

	var guids map[string]string

	if *newGUIDs {
		fmt.Println("Generating new guids...")

		guidList, err := c.extractGUIDsFromSln(slnPath)

		if err != nil {
			fmt.Printf("Failed to extract guids from the solution file: %v\n", err)
			return 1
		}

		guids = c.GeneratenewGUIDs(guidList)

		for k, v := range guids {
			fmt.Printf("  %s -> %s\n", k, v)
		}
	}

	err = copyItems(dirList, guids)

	if err != nil {
		fmt.Printf("Failed generate new items: %v\n", err)
		return 1
	}

	return 0
}

func getDirectoryList(dirPath string) (map[string]os.FileInfo, string, error) {
	dirList := make(map[string]os.FileInfo)
	slnPath := ""

	ignoredDirs := []string{}

	err := filepath.Walk(dirPath, func(filePath string, fi os.FileInfo, walkErr error) error {
		// if an error occurred while walking the directory bubble up
		if walkErr != nil {
			return walkErr
		}

		// get the file path relative to the dir path
		rel := strings.Replace(filePath, dirPath, "", 1)

		name := fi.Name()

		if len(rel) == 0 {
			// if we are at the root, or are looking at an ignored file skip it
			return nil
		} else if _, ok := dotnet.VsData.IgnoredItems[name]; ok {
			// we need to keep track of the directories we ignored so
			// that we do not try to copies files within the directory
			if fi.IsDir() {
				ignoredDirs = append(ignoredDirs, rel)
			}

			return nil
		}

		// check to see if this path begins with any of our ignore directories
		for _, ignoredDir := range ignoredDirs {
			if strings.HasPrefix(rel, ignoredDir) {
				return nil
			}
		}

		ext := strings.ToLower(filepath.Ext(name))

		if ext == ".sln" {
			slnPath = filePath
		}

		dirList[rel] = fi

		return nil
	})

	return dirList, slnPath, err
}

func (c *DotnetCommand) extractGUIDsFromSln(filePath string) (guids []string, err error) {
	contents, err := ioutil.ReadFile(filePath)
	if err != nil {
		return
	}

	guids, err = c.ExtractGUIDsFromString(string(contents))

	return
}

// ExtractGUIDsFromString extracts guids from file contents
func (c *DotnetCommand) ExtractGUIDsFromString(contents string) (guids []string, err error) {
	lines := strings.Split(contents, "\n")

	guids = []string{}

	for _, line := range lines {
		matches := guidInSln.FindStringSubmatch(strings.TrimSpace(line))

		if len(matches) > 1 {
			guids = append(guids, matches[1])
		}
	}

	return
}

// GeneratenewGUIDs takes a list of GUIDs and generates a new GUID to replaced
// each GUID in the list
func (c *DotnetCommand) GeneratenewGUIDs(guids []string) (guidMap map[string]string) {
	guidMap = make(map[string]string)

	for _, guid := range guids {
		guidMap[guid] = strings.ToUpper(uuid.NewV4().String())
	}

	return
}

func copyItems(dirList map[string]os.FileInfo, guids map[string]string) (err error) {
	fmt.Println("Generating new items...")

	// first we need to get our list of paths in order
	paths := []string{}

	for relPath := range dirList {
		paths = append(paths, relPath)
	}

	sort.Strings(paths)

	// now that we have the paths in order we can iteration and copy/transform the items
	for _, relPath := range paths {
		// transform the path to use the new solution name
		dstRel := strings.Replace(relPath, tplName, dstName, -1)

		tpl := path.Join(tplPath, relPath)
		dst := path.Join(dstPath, dstRel)

		err = copyItem(tpl, dst, dirList[relPath], guids)
		if err != nil {
			return err
		}

		fmt.Printf("  %s -> %s\n", relPath, dstRel)
	}

	return err
}

func copyItem(src, dst string, fi os.FileInfo, guids map[string]string) (err error) {
	if fi.IsDir() {
		if _, err := os.Stat(dst); os.IsNotExist(err) {
			os.Mkdir(dst, 0755)
		}
	} else {
		ext := strings.ToLower(filepath.Ext(fi.Name()))

		// if this file is a type known to contain project names,
		// edit it, otherwise just copy it
		if _, ok := dotnet.VsData.SourceFileExts[ext]; ok {
			err = transformFile(src, dst, guids)
		} else if _, ok := dotnet.VsData.ProjectFileExts[ext]; ok {
			err = transformFile(src, dst, guids)
		} else {
			err = copyFile(src, dst)
		}
	}

	return
}

func transformFile(src, dst string, guids map[string]string) (err error) {
	read, err := ioutil.ReadFile(src)
	if err != nil {
		return
	}

	// rename
	newContents := strings.Replace(string(read), tplName, dstName, -1)

	// replace guids
	if guids != nil {
		for k, v := range guids {
			newContents = strings.Replace(newContents, k, v, -1)
		}
	}

	err = ioutil.WriteFile(dst, []byte(newContents), 0666)

	return
}

// copyFile copies the contents of the file named src to the file named
// by dst. The file will be created if it does not already exist. If the
// destination file exists, all it's contents will be replaced by the contents
// of the source file.
func copyFile(src, dst string) (err error) {
	in, err := os.Open(src)
	if err != nil {
		return
	}
	defer in.Close()

	out, err := os.Create(dst)
	if err != nil {
		return
	}
	defer func() {
		cerr := out.Close()
		if err == nil {
			err = cerr
		}
	}()

	if _, err = io.Copy(out, in); err != nil {
		return
	}
	err = out.Sync()

	return
}

// Synopsis gives an overview of this command
func (c *DotnetCommand) Synopsis() string {
	return "Create a new .NET solution from a .NET solution"
}

// Help provides detailed usage information for this command
func (c *DotnetCommand) Help() string {
	helpText := `
usage: ouroboros dotnet <args> [tpl path] [dst path]
  
Available args are:  
    new-guids     Generate new project GUIDs
	tpl-name      Override template solution name
`
	return strings.TrimSpace(helpText)
}
